load-testing/initial-load-testing.md

```diff
@@ -0,0 +1,89 @@
+| jpg          | 10000           | 160b      | Failed<sup>6</sup>      | n/a                                                     | n/a                |
+
+
+<sup>1</sup>The upload time includes the time taken to upload the file and run the client side processing
+
+<sup>2</sup>The backend processing time is the total time the backend checks took to run so the time taken by the slowest of the checks.
+
+<sup>3</sup> The maximum size for a non multipart upload is 5Gb. Because this file was random bytes, gzip didn't compress it much so the consignment export size was over this limit.
+
+<sup>4</sup> The api was responding too slowly so the API client in the download files lambda timed out.
+
+<sup>5</sup> The majority of backend checks that did pass should have completed a lot faster because each check for an individual file should only take around 1 minute. This suggests that SQS is sending a lot of requests to each individual lambda which is increasing the time it takes.
+
+<sup>6</sup> The browser crashed while trying to send the client side processing results to the API. The response json from the client metadata call is around 50Mb so the browser is running out of memory. I tried this in Chrome and Firefox. I did get the upload to succeed by using a t3.2xlarge which has 8 cores and 32Gb of RAM but this is much larger than the expected machines the users will use though.
+
+## Problems found during the tests
+
+### Timeouts
+There are 4 timeouts that were preventing large consignments being processed.
+* The akka request timeout. This is the maximum time a request can take
+* The akka server timeout. The maximum time a server response can take. This needs to be the same as the request timeout as they basically do the same thing.
+* The load balancer idle timout.
+* The client timeout when contacting the API. When the API is overloaded, requests can take longer and the clients are timing out.
+
+### Logging
+There is little to no logging in the backend checks lambdas. The antivirus prints the current file id but the others only log errors. This makes it difficult to find out what has happened when there are a lot of files processed. Also, we get one line of a stack trace per cloudwatch event which makes them difficult to read.
+
+### Client metadata api response json
+We are returning a large json payload from the client metadata mutation which is consuming a lot of memory in the browser and causing the browser to crash on lower spec machines.
+
+### No mulipart uploads in export task
+We are using the single file upload call in the export task which has a size limit of 5Gb
+
+### Parallelism in the backend check lambdas
+It seems that AWS is passing a lot of messages to individual lambdas which is making them slow. We can reduce the number of messages passed to an individual lambda but this will increase the cost as more lambdas will run.
```

Comment:
Does this mean batching or something else? i.e. are you talking about reducing the number of messages by making the download lambda download multiple files?

Reply:
For most of the lambdas, we just let AWS decide how many messages to pass to the lamdba which means that each individual invocation can be processing 10 or more files which takes more time. We could do like we've done with the download files lambda https://github.com/nationalarchives/tdr-terraform-modules/blob/master/lambda/download_files.tf#L45 where we set the batch size to 1. This will reduce the time each individual lambda invocation takes to run but will cost us more money because we're invoking more lambdas.

Reply:
Ahh, I didn't realise the download files lambda was the only one with a batch size of one. Possibly something to tune when we start dealing with much larger consignments in production.